Switch from Spectral to Vacuum (#171)
ukhsa-collaboration/standards-api: docs/linting-rules/index.md

Run Vacuum against your OpenAPI definition, referencing the UKHSA ruleset. The custom rules rely on JavaScript functions, so build them first and pass the functions directory to Vacuum.

You must install the ruleset via npm and then reference it; the UKHSA ruleset npm package is hosted on GitHub — see GitHub's documentation [Installing a GitHub npm package][5].

The simplest way to authenticate against the Github NPM registry locally is to use the npm login command (as described [here][6]) to authenticate with GitHub Packages, this adds the required credentials to your npm configuration file i.e. `.npmrc`.

```sh
$ npm login --scope=@ukhsa-collaboration --auth-type=legacy --registry=https://npm.pkg.github.com

Username: USERNAME #GITHUB USERNAME
Password: TOKEN #GITHUB PAT TOKEN
```

```sh
npm install @ukhsa-collaboration/openapi-linting-rules --save-dev
```

Build the Vacuum-compatible functions (outputs to `dist-vacuum/functions`) and run the linter with the `--functions` flag so the custom rules load correctly.

```sh
npm run build:functions:cjs
npx vacuum lint openapi-definition.yml \
  -r ./node_modules/@ukhsa-collaboration/openapi-linting-rules/ukhsa.oas.rules.yml \

>   --functions ./node_modules/@ukhsa-collaboration/openapi-linting-rules/dist-vacuum/functions

Reviewer: `dist-vacuum`? is that vacuum foldername convention?

Author: Applied. Dropped the dual build and now use a single dist/ output. The docs/examples now point to dist/functions, and dist-vacuum is no longer used.

```

> If you want Vacuum to resolve every `$ref` before evaluating rules, add `--resolve-refs` (default off). To resolve nested `$ref` values relative to their own document context, add `--resolve-nested-refs`. The global `--resolve-refs` overrides any per-rule `resolveRefs` settings, while still honoring `--remote` and `--base` for external lookups.

Reviewer: is this stuff mandatory for our rulesets? if so we should make sure they are included in usage examples.

or do we make use of [vacuum config files](https://quobix.com/vacuum/configuring/#need-to-configure-vacuum-in-a-consistent-way)?

Author: Applied. Updated docs/linting-rules/index.md to use vacuum.conf.yaml in the main usage example and corrected the flag names for our current Vacuum version (--resolve-all-refs and --nested-refs-doc-context).

The docs now explicitly state these ref-resolution options are recommended (not mandatory), and removed the unsupported --jsonpath-lazy mention for Vacuum 0.26.5.

Where rules been adopted from from existing open source API rulesets a link is supplied on the relevant rule page.

## How to use the rules

### Install Vacuum

[Vacuum][3] is a fast OpenAPI linter for rulesets like this one. Install Vacuum globally or as a dev dependency.

```sh
npm install @quobix/vacuum --save-dev
```

Read the [official Vacuum documentation][4] for more installation options.

### Run Vacuum against your OpenAPI definition

Run Vacuum against your OpenAPI definition, referencing the UKHSA ruleset. The npm package ships prebuilt custom JavaScript functions, so most consumers only need to point Vacuum at the packaged `dist/functions` directory.

You must install the ruleset via npm and then reference it; the UKHSA ruleset npm package is hosted on GitHub — see GitHub's documentation [Installing a GitHub npm package][5].

> The simplest way to authenticate against the Github NPM registry locally is to use the npm login command (as described [here][6]) to authenticate with GitHub Packages, this adds the required credentials to your npm configuration file i.e. `.npmrc`.

Reviewer: can we be consistant when using the name `GitHub`  in docs, it should be a capital H

Author: Corrected

> [!TIP] Need a JSON report?
> Use `npx vacuum spectral-report <spec> <report.json>` to generate a report you can post-process in existing tooling.

### Review and fix any reported issues

Once the linter has highlighted any issues or errors, review and fix to ensure your OpenAPI definition remains compliant with the UKHSA guidelines.

### Severity overrides

Some OpenAPI definitions produced by platforms like [pygeoapi][7] might struggle to meet every **MUST** requirement. This is often because these issues are outside the consuming team's direct control, as fixes would require upstream contributions or maintaining a fork. To accommodate this, the pygeoapi ruleset relaxes the severity of certain rules when an OpenAPI document is identified as being generated by pygeoapi.

See [severity overrides][8] for more information.

> [!NOTE]
> Use the pygeoapi ruleset (`ukhsa.oas.rules.pygeoapi.yml`) or the wrapper script in `scripts/vacuum-lint.mjs` to select the correct ruleset based on `info.x-api-type`.

> ### CI/CD Github Actions

Reviewer: same here

Author: Applied for consistency - removed the extra note styling in that area and kept the guidance as normal instructional text

> The following is a sample Github actions job which can be used as an example of setting up linting as part of you CI/CD pipeline.

Reviewer: and here.

Author: Aligned this section wording/style with the rest of the page

<!-- {% raw %} -->

```yaml
...
---
name: API Standards Checks

on:
  pull_request:
    branches: [main, master] # Example only - you would typically use either 'main' or 'master', not both

jobs:
  lint-openapi:
    name: Lint OpenAPI
    runs-on: ubuntu-latest

    permissions:
      contents: read
      issues: read
      checks: write
      pull-requests: write
      packages: read

    steps:
      - name: Checkout code
        uses: actions/checkout@v5

      - uses: actions/setup-node@v6
        with:
          node-version: "22.x"
          registry-url: "https://npm.pkg.github.com"
          # Defaults to the user or organization that owns the workflow file
          scope: "@ukhsa-collaboration"

>       - run: npm install @ukhsa-collaboration/openapi-linting-rules

Reviewer: indentation seems wrong here, can we copy and paste the workflow into vscode and lint the workflow? make sure the format is correct.

```suggestion
        - name: Install OpenAPI Linting Rules
          run: npm install @ukhsa-collaboration/openapi-linting-rules
```

Author: Updated

resolve-all-refs: true
nested-refs-doc-context: true
```

<!-- {% raw %} -->

```yaml
name: API Standards Checks

on:
  pull_request:
    branches: [main, master] # Example only - you would typically use either 'main' or 'master', not both

jobs:
  lint-openapi:
    name: Lint OpenAPI
    runs-on: ubuntu-latest

    permissions:
      contents: read
      issues: read
      checks: write
      pull-requests: write
      packages: read

    steps:
      - name: Checkout code
        uses: actions/checkout@v7

      - name: Set up Node.js
        uses: actions/setup-node@v6
        with:
          node-version: "22.x"
          registry-url: "https://npm.pkg.github.com"
          # Defaults to the user or organization that owns the workflow file
          scope: "@ukhsa-collaboration"

      - name: Install linting dependencies
        run: npm install @quobix/vacuum @ukhsa-collaboration/openapi-linting-rules
        env:
          NODE_AUTH_TOKEN: ${{ secrets.GITHUB_TOKEN }}

      - name: Lint OpenAPI specifications
        run: |
          # Uses the vacuum.conf.yaml committed to the repository root.

>           npx vacuum lint ./*.{json,yml,yaml}

Reviewer: ```suggestion
          npx vacuum lint ./*.{json,yml,yaml}
          # Example: Only lint OpenAPI files
          # npx vacuum lint ./@(openapi|swagger|*api)*.{json,yml,yaml}
```

Author: Added back into the main example. I also kept the shopt -s extglob line alongside it so the selective pattern is documented in a form that actually works when someone uncomments it.

```

<!-- {% endraw %} -->

The unquoted shell pattern above relies on shell expansion to target JSON and YAML files in the current working directory before Vacuum runs.

If you want to restrict matching to filenames that start with `openapi`, `swagger`, or contain `api`, use a bash step with extended globbing enabled:

```yaml
      - name: Lint OpenAPI specifications with stricter filename matching
        shell: bash
        run: |
          shopt -s extglob
          npx vacuum lint ./@(openapi|swagger|*api)*.{json,yml,yaml}
```

> 

Reviewer: ```suggestion
The above example uses [glob syntax][13] to target only OpenAPI specification files.

The glob pattern `./@(openapi|swagger|*api)*.{json,yml,yaml}` matches:

> [!TIP] Matches
>
> ```text
> openapi.json
> something.api.yaml
> swagger.json
> ```

The global pattern does not match:

> [!CAUTION] Does Not Match
>
> ```text
> /node_modules/openapi.yaml
> /.git/something.json
> ```
```

Did we mean to remove this?

Author: Yes, we did mean to keep that explanatory block. I restored the stricter-pattern explanation and the match / does-not-match examples, and kept an explicit note that bash extended globbing needs to be enabled for that selective pattern.

---
caption: API Design Guidelines
includeInBreadcrumbs: true
eleventyNavigation:
  parent: api-design-guidelines
---

# Linting Rules

## Overview

A linting ruleset was created to support API Developers/Providers in achieving the standards described in the [UKHSA API Guidelines][1], ensuring consistency, reliability, and security across all APIs developed within or on behalf of UKHSA.

> As well as the rules described herein, the UKHSA ruleset extends Vacuum's built-in OpenAPI recommended baseline (`vacuum:oas`). These common sense rules help ensure an OpenAPI definition adheres to the [OpenAPI specification][2], and include documentation-focused checks that encourage high quality, rich API descriptions for the best possible APIM Developer Portal experience.

Reviewer: we should probably link out to the rules here:
https://quobix.com/vacuum/rules/

Author: Added a direct link to the Vacuum built-in rules reference in the overview section.

As well as the rules described herein, the UKHSA ruleset extends Vacuum's built-in OpenAPI recommended baseline (`vacuum:oas`). These common sense rules help ensure an OpenAPI definition adheres to the [OpenAPI specification][2], and include documentation-focused checks that encourage high quality, rich API descriptions for the best possible APIM Developer Portal experience.

Where rules have been adopted from existing open source API rulesets, a link is supplied on the relevant rule page.

## Migration from Spectral to Vacuum

> The UKHSA ruleset previously supported Spectral-oriented consumption. It is now distributed for Vacuum consumption only.

Reviewer: do we need explain the reational for moving to vacuum?

Author: Added

> For users of `@ukhsa-collaboration/openapi-linting-rules`, this is a breaking tooling change: Spectral commands and Spectral configuration files are no longer the supported integration path for current releases of this package.

Reviewer: the existing package is `@ukhsa-collaboration/spectral-rules` to migrate to the new vacuum based linting the consumers need to switch to `@ukhsa-collaboration/openapi-linting-rules`

Author: Thanks, updated. The migration section now explicitly states that Spectral users move from @ukhsa-collaboration/spectral-rules to @ukhsa-collaboration/openapi-linting-rules, and marks spectral-rules as legacy.

> There is no separate replacement Spectral package for new releases. The migration path is to keep using this same UKHSA package with Vacuum.

Reviewer: technically this is a new package when migrating to vacuum right?

Author: Yes, agreed. I reworded the migration text to make the package transition explicit (legacy Spectral package to Vacuum-first package), instead of implying in-place tooling continuity.

### What existing Spectral users should do next

1. Install Vacuum in your project (`npm install @quobix/vacuum --save-dev`).

> 2. Install or update the UKHSA ruleset package (`npm install @ukhsa-collaboration/openapi-linting-rules --save-dev`).

Reviewer: - uninstall `@ukhsa-collaboration/spectral-rules`

also dont forget to run the markdown fixer as I think it prefers numbered lists to all be a number 1.? from memory

Author: Done. I added an explicit uninstall step for @ukhsa-collaboration/spectral-rules and normalized the ordered list numbering style. I also ran the markdown formatter/linter on the updated docs.

3. Replace Spectral config with a `vacuum.conf.yaml` file that points to `ukhsa.oas.rules.yml` and `dist/functions`.

> 4. Replace Spectral lint commands in local scripts and CI jobs with `npx vacuum lint ... --config ./vacuum.conf.yaml`.

Reviewer: or execute via the wrapper which detects api type ?

Author: Added as an explicit optional step in migration guidance for this repository, pointing to vacuum-lint.mjs for auto-selection of pygeoapi override behavior.

The UKHSA ruleset previously supported Spectral-oriented consumption via `@ukhsa-collaboration/spectral-rules`.

Current releases are distributed for Vacuum consumption via `@ukhsa-collaboration/openapi-linting-rules`. This is a breaking tooling change for teams moving from the Spectral package.

`@ukhsa-collaboration/spectral-rules` should be treated as a legacy package and should not be used for new integrations.

### Why we moved to Vacuum

- Faster linting in local development and CI pipelines.
- Improved handling of very large OpenAPI documents, including pygeoapi-generated specifications.

### What existing Spectral users should do next

1. Install Vacuum in your project (`npm install @quobix/vacuum --save-dev`).
1. Uninstall the legacy Spectral package (`npm uninstall @ukhsa-collaboration/spectral-rules`).
1. Install or update the Vacuum-first UKHSA ruleset package (`npm install @ukhsa-collaboration/openapi-linting-rules --save-dev`).
1. Replace Spectral config with a `vacuum.conf.yaml` file that points to `ukhsa.oas.rules.yml` and `dist/functions`.
1. Replace Spectral lint commands in local scripts and CI jobs with `npx vacuum lint ... --config ./vacuum.conf.yaml`.

> 1. For this repository, you can alternatively run `node scripts/vacuum-lint.mjs` to auto-select the pygeoapi override ruleset when needed.

Reviewer: would like to discuss this bit - is this script not distrubuted in the npm package? it would be useful for the apim onboarding repo.

Author: You are right, it is not distributed, should we
- Keep it repo-only - update docs to clearly say this is only for this repo and give external repos a supported equivalent command/path.

- Make it supported for consumers - ship a public wrapper (for example a package bin command) so onboarding repos can use it directly
?

Reviewer: I think we need to make it available to be used directly so teams have everything they need.

Author: The wrapper is now shipped in the npm package as a supported command (ukhsa-vacuum-lint), and docs were updated to point consumers to that command rather than a repo-local script path.

### Run Vacuum against your OpenAPI definition

> Run Vacuum against your OpenAPI definition, referencing the UKHSA ruleset. The npm package ships prebuilt custom JavaScript functions, so most consumers only need to point Vacuum at the packaged `dist/functions` directory.

Reviewer: ```suggestion
Run Vacuum against your OpenAPI definition, referencing the UKHSA ruleset.
```
I'm not sure the rest of the paragraph adds value - thoughts?

Author: Simplified this paragraph

npm install @ukhsa-collaboration/openapi-linting-rules --save-dev
```

> Create a `vacuum.conf.yaml` file so Vacuum can load the UKHSA ruleset and custom functions by default.

Reviewer: ```suggestion
create a local vacuum configuration file `vacuum.conf.yaml` to pre-configure the use of the UKHSA ruleset.
```

Author: Applied

```yaml
ruleset: ./node_modules/@ukhsa-collaboration/openapi-linting-rules/ukhsa.oas.rules.yml

> functions: ./node_modules/@ukhsa-collaboration/openapi-linting-rules/dist/functions

Reviewer: arent these functions already referenced (relatively) in `ukhsa.oas.rules.yml`?

Author: Keeping this as Vacuum (current 0.26.x) does not load custom functions from functionsDir declared inside a ruleset. Consumers do need to tell Vacuum where the functions live via --functions (CLI) or functions: (in vacuum.conf.yaml).

resolve-all-refs: true
nested-refs-doc-context: true

> ```

Reviewer: we should probably provide a exemplar configuration example for people to copy paste in fact we could provide an echo command like the original spectral instructions?

```sh
cat << 'EOF' > vacuum.conf.yaml
ruleset: ./node_modules/@ukhsa-collaboration/openapi-linting-rules/ukhsa.oas.rules.yml
functions: ./node_modules/@ukhsa-collaboration/openapi-linting-rules/dist/functionsCollapse 
resolve-all-refs: true
nested-refs-doc-context: true
EOF
```

Author: Added

Then run Vacuum against your OpenAPI definition.

```sh
npx vacuum lint ./path/to/openapi-definition.yml \
  --config ./vacuum.conf.yaml
```

> [!NOTE]

> > `./path/to/openapi-definition.yml` is a placeholder. Replace it with the real path to your OpenAPI file.

Reviewer: I think the placeholder is obvious in the example so this note is not really needed.

Author: Removed

> [!NOTE]

> > Only run `npm run build:functions:cjs` when developing this ruleset repository locally and changing custom function source files. For normal use of the published npm package, that build step is not required.

Reviewer: this doesnt live in the standard docs IMO - its more something that lives in CONTRIBUTING.md or README.md as its more related to the development of the rulesets. - lets keep the docs instructions focused on using the rules. - we can keep the other stuff but it doesnt live here.

Author: Removed development-only build guidance from the linting docs

> `resolve-all-refs` and `nested-refs-doc-context` are recommended for this ruleset (several custom rules expect refs to be resolved), but they are not strictly mandatory. You can remove either option from `vacuum.conf.yaml` or override them on the command line when needed.

Reviewer: if our rules expect refs then we should make sure people have it configured correctly the way we need it. lets not tell them all the cli options of vacuum but focus on the minimum required to use the ukhsa ruleset

Author: Removed the optional CLI-detail section and documented only the minimum required config for this ruleset

### Run Vacuum against your OpenAPI definition

Run Vacuum against your OpenAPI definition, referencing the UKHSA ruleset.

You must install the ruleset via npm and then reference it; the UKHSA ruleset npm package is hosted on GitHub — see GitHub's documentation [Installing a GitHub npm package][5].

The simplest way to authenticate against the GitHub NPM registry locally is to use the npm login command (as described [here][6]) to authenticate with GitHub Packages, this adds the required credentials to your npm configuration file i.e. `.npmrc`.

```sh
$ npm login --scope=@ukhsa-collaboration --auth-type=legacy --registry=https://npm.pkg.github.com

> Username: USERNAME #GitHub username

Reviewer: these could have remained as was
```suggestion
Username: USERNAME #GITHUB USERNAME
Password: TOKEN #GITHUB PAT TOKEN
```

Password: TOKEN #GitHub PAT token
```

```sh
npm install @ukhsa-collaboration/openapi-linting-rules --save-dev
```

Create a local vacuum configuration file `vacuum.conf.yaml` to pre-configure the use of the UKHSA ruleset.

```sh
cat << 'EOF' > vacuum.conf.yaml
ruleset: ./node_modules/@ukhsa-collaboration/openapi-linting-rules/ukhsa.oas.rules.yml
functions: ./node_modules/@ukhsa-collaboration/openapi-linting-rules/dist/functions
resolve-all-refs: true
nested-refs-doc-context: true
EOF
```

Then run Vacuum against your OpenAPI definition.

```sh

> npx vacuum lint ./path/to/openapi-definition.yml \

Reviewer: I thought the config file is picked up by convention? no need to specify it? or do you need to be explicit

Author: Good catch - confirmed, Vacuum auto-discovers vacuum.conf.yaml from the working directory (`--config string config file (defaults to vacuum.conf.yaml)`). I dropped the explicit `--config ./vacuum.conf.yaml` from the CLI examples in index.md (and CONTRIBUTING.md for consistency) and added a one-line note that it is auto-discovered.

```sh
$ npm login --scope=@ukhsa-collaboration --auth-type=legacy --registry=https://npm.pkg.github.com

Username: USERNAME #GITHUB USERNAME
Password: TOKEN #GITHUB PAT TOKEN

> Password: TOKEN #GitHub PAT token

Reviewer: duplicate line
```suggestion
```

Author: Done

          run: |

>             npx vacuum lint "*.{json,yml,yaml}" \

Reviewer: if we are incurraging the use of config file then perhaps we could simplify this ? thoughts?

Author: Agreed and applied. The CI lint command now uses the config file directly, without repeating ruleset/functions flags

    steps:
      - name: Checkout code
        uses: actions/checkout@v7

      - name: Set up Node.js
        uses: actions/setup-node@v6
        with:
          node-version: "22.x"
          registry-url: "https://npm.pkg.github.com"
          # Defaults to the user or organization that owns the workflow file
          scope: "@ukhsa-collaboration"

      - name: Install linting dependencies
        run: npm install @quobix/vacuum @ukhsa-collaboration/openapi-linting-rules
        env:
          NODE_AUTH_TOKEN: ${{ secrets.GITHUB_TOKEN }}

>       - name: Create Vacuum config

Reviewer: I dont think we should advise teams to create this config on the fly - it should probably be commited to the target repositories by the teams.

Author: Agree - teams should commit vacuum.conf.yaml to their repo root so it's reused by both local dev and CI rather than generated on the fly. I removed the `Create Vacuum config` step from the GitHub Actions example and added a short note (plus a copy-pasteable snippet of the file's contents) telling teams to commit vacuum.conf.yaml at the repository root.

          # Defaults to the user or organization that owns the workflow file
          scope: "@ukhsa-collaboration"

      - name: Install linting dependencies
        run: npm install @quobix/vacuum @ukhsa-collaboration/openapi-linting-rules
        env:
          NODE_AUTH_TOKEN: ${{ secrets.GITHUB_TOKEN }}

      - name: Lint OpenAPI specifications
        run: |
          npx vacuum lint "*.{json,yml,yaml}" \

>             -r ${{ github.workspace }}/node_modules/@ukhsa-collaboration/openapi-linting-rules/ukhsa.oas.rules.yml \

Reviewer: this sample should probably be based on a vacuum config and thus could be simplified (no need to pass -r and --functions args? but with perhaps a guard that checks for the config? or at the very least a comment that points to the required config file?

Author: Aligned

<!-- {% endraw %} -->

The above example uses [glob syntax][10] to target only OpenAPI specification files.

> The glob pattern `@(openapi|swagger|*api)*.{json,yml,yaml}` matches:

Reviewer: you might need to update this if the you need to update the glob pattern.

Author: Updated
